File: DIP/FuelGauge.ts
// Abstraction
interface FuelGauge {
    level(): number;
}

// Real implementation
class RemoteFuelGauge implements FuelGauge {
    level(): number {
        return 80; // Real hardware value
    }
}

// Test implementation
class TestingFuelGauge implements FuelGauge {
    level(): number {
        return 100; // Test value
    }
}

// Higher level module
class RocketLauncher {
    private fuelGauge: FuelGauge;

    constructor(fuelGauge: FuelGauge) {
        this.fuelGauge = fuelGauge;
    }

    isSafeToLaunch(): boolean {
        return this.fuelGauge.level() > 50;
    }
}

// Test
const testGauge = new TestingFuelGauge();
const rocket = new RocketLauncher(testGauge);

console.log(rocket.isSafeToLaunch());

File: DIP/FuelGauge.cs
// Abstraction
public interface IFuelGauge
{
    int Level();
}

// Real implementation
public class RemoteFuelGauge : IFuelGauge
{
    public int Level()
    {
        return 80; // Real hardware value
    }
}

// Test implementation
public class TestingFuelGauge : IFuelGauge
{
    public int Level()
    {
        return 100; // test value
    }
}

class RocketLauncher
{
    private IFuelGauge _fuelGauge;

    public RocketLauncher(IFuelGauge fuelGauge)
    {
        _fuelGauge = fuelGauge;
    }

    public bool IsSafeToLaunch()
    {
        return _fuelGauge.Level() > 50;
    }
}

// Test
class Program
{
    static void Main()
    {
        IFuelGauge testGauge = new TestingFuelGauge();
        RocketLauncher rocket = new RocketLauncher(testGauge);

        Console.WriteLine(rocket.IsSafeToLaunch());
    }
}
